### PR TITLE
Implements dce + hmd devices

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**
All credit goes to @mikusp for reaching this.

### CODE REVIEW

File: src/kernel/src/main.rs [hunk 1]
```diff
@@ -650,6 +657,15 @@ enum KernelError {
     #[error("dmem manager initialization failed")]
     DmemManagerInitFailed(#[from] DmemManagerInitError),
 
+    #[error("sbl_srv manager initialization failed")]
+    SblSrvManagerInitFailed(#[from] SblSrvInitError),
```
Comment (ultimaweapon) [inline src/kernel/src/main.rs:661]: Why `from` here?

File: src/kernel/src/main.rs [hunk 2]
```diff
@@ -650,6 +657,15 @@ enum KernelError {
...
+
+    #[error("hmd manager initialization failed")]
+    HmdManagerInitFailed(#[from] HmdInitError),
```
Comment (ultimaweapon) [inline src/kernel/src/main.rs:664]: Same here.

File: src/kernel/src/main.rs [hunk 3]
```diff
@@ -650,6 +657,15 @@ enum KernelError {
...
+
+    #[error("dce manager initialization failed")]
+    DceManagerInitFailed(#[from] DceInitError),
```
Comment (ultimaweapon) [inline src/kernel/src/main.rs:667]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by SuchAFuriousDeath** [comment]
Here is his experimental branch by the way: https://github.com/mikusp/obliteration/tree/wip_current.
It seems that he's reaching game code that draws to the screen for some games.
image

File: src/kernel/src/dev/sbl_srv.rs [hunk 4]
```diff
@@ -19,3 +30,32 @@ impl DeviceDriver for SblSrv {
...
+
+pub struct SblSrvManager {
+    sbl: Arc<CharacterDevice>,
+}
+
+impl SblSrvManager {
+    pub fn new() -> Result<Arc<Self>, SblSrvInitError> {
+        let sbl = make_dev(
+            SblSrv {},
+            DriverFlags::from_bits_retain(0x80000004),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/sbl_srv.rs:42]: Where did you get this value from?

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/dev/sbl_srv.rs (hunk 4 above)
Comment (SuchAFuriousDeath) [reply src/kernel/src/dev/sbl_srv.rs]: From the branch. But it seems it was wrong

**Comment by SuchAFuriousDeath** [comment]
It seems that the open functions have been left out as well.

File: src/kernel/src/dev/sbl_srv.rs [hunk 5]
```diff
@@ -19,3 +30,32 @@ impl DeviceDriver for SblSrv {
...
+impl SblSrvManager {
+    pub fn new() -> Result<Arc<Self>, SblSrvInitError> {
+        let sbl = make_dev(
+            SblSrv {},
+            DriverFlags::D_INIT,
+            0,
+            "sbl_srv",
+            Uid::ROOT,
+            Gid::ROOT,
+            Mode::new(0o444).unwrap(),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/sbl_srv.rs:47]: This also incorrect.

File: src/kernel/src/dev/sbl_srv.rs [hunk 6]
```diff
@@ -19,3 +30,32 @@ impl DeviceDriver for SblSrv {
...
+            None,
+            MakeDevFlags::ETERNAL,
```
Comment (ultimaweapon) [inline src/kernel/src/dev/sbl_srv.rs:49]: Same here.

File: src/kernel/src/dev/sbl_srv.rs [hunk 7]
```diff
@@ -19,3 +30,32 @@ impl DeviceDriver for SblSrv {
...
+
+        Ok(Arc::new(Self { sbl }))
+    }
+}
+
+/// Represents an error when [`SblSrvManager`] fails to initialize.
+#[derive(Debug, Error)]
+pub enum SblSrvInitError {
+    #[error("cannot create sbl_srv device")]
+    CreateSblSrvFailed(#[from] MakeDevError),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/sbl_srv.rs:60]: Why `from` here?

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/dev/hmd.rs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            Gid::ROOT,
+            Mode::new(0o666).unwrap(),
+            None,
+            MakeDevFlags::empty(),
+        )
+        .map_err(HmdInitError::CreateHmdCmdFailed)?;
+
+        let hmd_snsr = make_dev(
+            HmdSnsr {},
+            DriverFlags::D_INIT,
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:109]: This is not correct.

File: src/kernel/src/dev/hmd.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            hmd_dist,
+        }))
+    }
+}
+
+/// Represents an error when [`HmdManager`] fails to initialize.
+#[derive(Debug, Error)]
+pub enum HmdInitError {
+    #[error("cannot create hmd_cmd device")]
+    CreateHmdCmdFailed(MakeDevError),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:159]: Why omit `source` attribute here?

File: src/kernel/src/dev/hmd.rs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    #[error("cannot create hmd_snsr device")]
+    CreateHmdSnsrFailed(MakeDevError),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:161]: Same here.

File: src/kernel/src/dev/hmd.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    #[error("cannot create hmd_3da device")]
+    CreateHmd3daFailed(MakeDevError),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:163]: Same here.

File: src/kernel/src/dev/hmd.rs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    #[error("cannot create hmd_dist device")]
+    CreateHmdDistFailed(MakeDevError),
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:165]: Same here.

File: src/kernel/src/dev/hmd.rs
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:1]: Why omit `hmd_cr` device?

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/dev/hmd.rs [hunk 13]
```diff
@@ -106,7 +133,7 @@ impl HmdManager {
 
         let hmd_snsr = make_dev(
             HmdSnsr {},
-            DriverFlags::D_INIT,
+            DriverFlags::INIT | DriverFlags::NEEDMINOR,
```
Comment (ultimaweapon) [inline src/kernel/src/dev/hmd.rs:136]: I don't think `NEEDMINOR` is a correct one.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by SuchAFuriousDeath** [comment]
By the way: 
image

**Comment by ultimaweapon** [comment]
Seems like it is the right decision to go with the VM to support multiple processes.

**Comment by ultimaweapon** [comment]
Another solution is we can fake those communication if it is not fragile or too complicated.

**Comment by SuchAFuriousDeath** [comment]
> Seems like it is the right decision to go with the VM to support multiple processes.

image

**Comment by ultimaweapon** [comment]
The good news is now I have full access to my PS4. This will be much easier to look around to learn how it works.

File: src/kernel/src/dev/gc.rs [hunk 14]
```diff
@@ -35,15 +37,29 @@ impl DeviceDriver for Gc {
...
-        _: Option<&VThread>,
+        td: Option<&VThread>,
     ) -> Result<(), Box<dyn Errno>> {
+        if self.suspended {
+            todo!("gc suspended")
+        }
+
+        let td = td.unwrap();
+
+        let gc_check_passed = td.proc().cred().unk_gc_check();
```
Comment (ultimaweapon) [inline src/kernel/src/dev/gc.rs:48]: The PS4 check thread credential, not process credential.

Comment (ultimaweapon) [inline src/kernel/src/dev/gc.rs:48]: I saw there is a call to a function after this.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/dev/gc.rs (hunk 14 above)
Comment (SuchAFuriousDeath) [reply src/kernel/src/dev/gc.rs:48]: That's devfs_get_cdevpriv (a FreeBSD function). I added a todo comment for now.

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**